app/controllers/katello/api/registry/registry_proxies_controller.rb

```diff
@@ -85,6 +84,223 @@ def registry_authorize
+      org_id = props[:organization]
+      unless org_id.present? && org_id == org_id.to_i.to_s
+        return render_podman_error(
+          "NAME_INVALID",
+          "Invalid format. Organization id must be an integer without leading zeros.",
+          :bad_request
+        )
+      end
+      @organization = Organization.find(org_id.to_i)
+      if @organization.nil?
+        return render_podman_error(
+          "NAME_UNKNOWN",
+          "Organization id not found: '#{org_id}'",
+          :not_found
+        )
+      end
+      true
+    end
+
+    def check_blob_push_product_label(props)
+      prod_label = props[:product]
+      unless prod_label.present? && prod_label.length > 0
+        return render_podman_error(
+          "NAME_INVALID",
+          "Invalid format. Product label cannot be blank.",
+          :bad_request
+        )
+      end
+      product = @organization.products.where("LOWER(label) = '#{prod_label}'") # convert to lowercase
+      # reject ambiguous products (possible due to lowercase conversion)
+      if product.length > 1
+        return render_podman_error(
+          "NAME_INVALID",
+          "Product label '#{prod_label}' is ambiguous. Try using an id-based container name.",
+          :conflict
+        )
+      end
+      if product.length == 0
+        return render_podman_error(
+          "NAME_UNKNOWN",
+          "Product not found: '#{prod_label}'",
+          :not_found
+        )
+      end
+      @product = product.first
+      true
+    end
+
+    def check_blob_push_product_id(props)
+      prod_id = props[:product]
+      unless prod_id.present? && prod_id == prod_id.to_i.to_s
+        return render_podman_error(
+          "NAME_INVALID",
+          "Invalid format. Product id must be an integer without leading zeros.",
+          :bad_request
+        )
+      end
+      @product = @organization.products.find(prod_id.to_i)
+      if @product.nil?
+        return render_podman_error(
+          "NAME_UNKNOWN",
+          "Product id not found: '#{prod_id}'",
+          :not_found
+        )
+      end
+      true
+    end
+
+    def check_blob_push_container(props)
+      # All incorrect params should have been rejected by this point, assume ok
+      @container_name = props[:name]
+      @container_path_input = "#{props[:organization]}/#{props[:product]}/#{props[:name]}"
+      @container_path_actual = "#{@organization.label}/#{@product.label}/#{@container_name}"
+      return true
+    end
+
+    def create_container_repo_if_needed
+      if @product.root_repositories.where(label: @container_name).empty?
+        root = @product.add_repo(
+          name: @container_name,
+          label: @container_name,
+          download_policy: 'immediate',
+          content_type: Repository::DOCKER_TYPE,
+          unprotected: true,
+          is_container_push: true
+        )
+        sync_task(::Actions::Katello::Repository::CreateRoot, root, @container_path_input)
```

Comment:
I found an interesting issue here -- there seems to be a race condition.

Pushing content is concurrent. That means that multiple threads could be in this if statement at once time. So, `::Actions::Katello::Repository::CreateRoot` is being run even when `@product.root_repositories.where(label: @container_name).empty?` turns false.

You can reproduce this in the console by sticking a debugger right before the sync_task call. Multiple different debuggers will run at once due to the concurrency, and they'll all queue up before the sync_task call.

This is tricky because we need a way for one of the concurrent calls from podman to become the "chosen one" and for the rest of the calls to wait. One idea that comes to mind would be to implement some sort of lock that is saved in the database.

Reply:
I confirmed that using a single-threaded development environment "solves" the issue, so concurrency does seem to be the problem.

Reply:
One idea would be to use https://apidock.com/rails/v6.0.0/ActiveRecord/Relation/create_or_find_by

(or maybe `create_or_find_by!`)

I wonder if it would be easier to stop using the CreateRoot action and instead create the database entries manually, especially since we already have to add container push conditionals to those actions.

Reply:
Just for inspiration, https://api.rubyonrails.org/v7.1.3.2/classes/ActiveRecord/Relation.html#method-i-find_or_initialize_by is also available -- it would allow to instantiate the RootRepository object OR actually get a record if it exists.

Reply:
> One idea would be to use https://apidock.com/rails/v6.0.0/ActiveRecord/Relation/create_or_find_by
> 
> (or maybe `create_or_find_by!`)
> 
> I wonder if it would be easier to stop using the CreateRoot action and instead create the database entries manually. We'd then be able to remove the conditionals from the current actions.

I've been experimenting with that for testing different approaches. I'm going to try and roll everything back into the foreman tasks unless that breaks concurrency, just so there isn't two different methods in Katello that do the same thing.

Reply:
Also thank you for the resources. Have been trying to get create_or_find_by to work but there are a lot of layers of foreman tasks to unroll.

Reply:
As of the newest push, I believe this is working correctly. I used `create_or_find_by` inside of `app/models/katello/glue/pulp/repos.rb`. When you have time, please confirm the issue is fixed on your end (I've been having trouble getting the endpoint to be hit simultaneously on my device).